File: ExpAddOperators.java
package BackTracking;

import java.util.ArrayList;
import java.util.List;

// Time Complexity : O(4^n)
// Space Complexity : O(h) h-> depth of recursive stack
// Did this code successfully run on Leetcode : Yes

public class ExpAddOperators {

    List<String> result;
    public List<String> addOperators(String num, int target) {

        result = new ArrayList<>();
        helper(num, target, new StringBuilder(),0, 0L,0L);
        return result;
    }

    public void helper(String num, int target, StringBuilder path, int pivot, Long tail, Long calc)
    {
        if(pivot == num.length() && calc == target)
        {
            result.add(new String(path));
            return;
        }

        for(int i=pivot; i<num.length(); i++)
        {
            Long curr = Long.parseLong(num.substring(pivot,i+1));
            int le = path.toString().length();

            //handle preceeding zeros
            if(i!=pivot && num.charAt(pivot)=='0') continue;

            if(pivot == 0)
            {
                path.append(curr);
                helper(num, target, path, i+1, curr, curr);
                path.setLength(le);
            }else{

                //+ operator
                path.append("+");
                path.append(curr);
                helper(num, target, path, i+1, curr, calc+curr);
                path.setLength(le);

                //-operator
                path.append("-");
                path.append(curr);
                helper(num, target, path,i+1, -curr, calc-curr);
                path.setLength(le);


                //* operator
                path.append("*");
                path.append(curr);
                helper(num, target, path,i+1, tail*curr, calc-tail+(tail*curr));
                path.setLength(le);


            }
        }

    }
}

File: Exp_add_operators.py
# TC: O(N X 4^N)
# SC: O(N X 4^N)

class Solution:
    res=[]
    def helper(self,num,target,pivot,path,calc,tail):
        if pivot==len(num) and target==calc:
            self.res.append(path)
            return

        for i in range(pivot,len(num)):
            if len(str(int(num[pivot:i+1])))==len(num[pivot:i+1]):
                cur=int(num[pivot:i+1])
                if pivot==0:
                    self.helper(num,target,i+1,str(cur),cur,cur)
                else:
                    self.helper(num,target,i+1,path+"+"+str(cur),calc+cur,cur)

                    self.helper(num,target,i+1,path+"-"+str(cur),calc-cur,-cur)

                    self.helper(num,target,i+1,path+"*"+str(cur),(calc-tail)+(tail*cur),tail*cur)

    def addOperators(self, num: str, target: int) -> List[str]:
        self.res=[]
        self.helper(num,target,0,"",0,0)
        return self.res



        